rtl/src/main/scala/tensil/mutex/LockPool.scala

```diff
@@ -0,0 +1,145 @@
+/* SPDX-License-Identifier: Apache-2.0 */
+/* Copyright © 2019-2022 Tensil AI Company */
+
+package tensil.mutex
+
+import chisel3._
+import chisel3.experimental.FixedPoint
+import chisel3.experimental.BundleLiterals._
+import chisel3.util.Decoupled
+import tensil.util.decoupled.Counter
+import chisel3.util.Queue
+import chisel3.util.DecoupledIO
+import tensil.util.zero
+import chisel3.util.log2Ceil
+
+class LockPool[T <: Data with Comparable[T]](
+    gen: T,
+    numActors: Int,
+    numLocks: Int,
+    select: T => UInt
+) extends Module {
+  val maxDelay = 1 << 4
+  val io = IO(new Bundle {
+    val actor = Vec(
+      numActors,
+      new Bundle {
+        val in  = Flipped(Decoupled(gen))
+        val out = Decoupled(gen)
+      }
+    )
+    val lock = Flipped(
+      Decoupled(
+        new ConditionalReleaseLockControl(
+          gen,
+          numActors,
+          numLocks,
+          maxDelay
+        )
+      )
+    )
+    val locked = Decoupled(
+      new ConditionalReleaseLockControl(
+        gen,
+        numActors,
+        numLocks,
+        maxDelay
+      )
+    )
+    val deadlocked = Decoupled(Bool())
+  })
+
+  val actor       = VecInit(io.actor.map(a => a.in))
+  val lockControl = io.lock
+
+  val lock = RegInit(
+    VecInit(
+      Array.fill(numLocks)(
+        zero(new ConditionalReleaseLock(gen, numActors, maxDelay))
+      )
+    )
+  )
+
+  val block = VecInit(for ((a, id) <- actor.zipWithIndex) yield {
+    // block by default
+    io.actor(id).out.noenq()
+    a.ready := false.B
+    val requiredLockId = select(a.bits)
+    val requiredLock   = lock(requiredLockId)
+    val blocked =
+      (requiredLock.held && requiredLock.by =/= id.U) || (lockControl.fire && lockControl.bits.lock === requiredLockId && lockControl.bits.by =/= id.U)
+    when(!blocked) {
+      // allow actor to proceed when not blocked
+      io.actor(id).out <> a
+    }
+    blocked
+  })
+
+  // locked output facilitates testing - not required to be connected in production
+  io.locked.bits <> lockControl.bits
+  io.locked.valid := lockControl.fire
+
+  // for signalling when deadlocked
+  io.deadlocked.noenq()
+
+  def acquire(l: ConditionalReleaseLock[T]): Unit = {
+    when(lockControl.valid) {
+      l.held := lockControl.bits.acquire
+      l.by := lockControl.bits.by
+      l.cond <> lockControl.bits.cond
+    }
+  }
+  def release(l: ConditionalReleaseLock[T]): Unit = {
+    l.held := false.B
+  }
+  // this covers the case where the condition happens to be observed on the same
+  // cycle the lock would have been acquired
+  val incomingObserved =
+    lockControl.valid && actor(lockControl.bits.by).fire && actor(
+      lockControl.bits.by
+    ).bits === lockControl.bits.cond
```

Comment:
Do we have test case where incomingObserved is true?

Reply:
Added a test covering this case to `LockPoolSpec`